src/plugins/dropdown/dropdown.js

```diff
@@ -0,0 +1,369 @@
+
+		dropdown.addEventListener( "focusin", this.onFocusin.bind( this ) );
+		dropdown.addEventListener( "focusout", this.onFocusout.bind( this ) );
+		window.addEventListener( "mousedown", this.onBackgroundMousedown.bind( this ), true );
+	}
+
+	setFocusToMenuitem( newMenuitem ) {
+		this.menuitemNodes.forEach( function( item ) {
+			if ( item === newMenuitem ) {
+				item.tabIndex = 0;
+				newMenuitem.focus();
+			} else {
+				item.tabIndex = -1;
+			}
+		} );
+	}
+
+	setFocusToFirstMenuitem() {
+		this.setFocusToMenuitem( this.firstMenuitem );
+	}
+
+	setFocusToLastMenuitem() {
+		this.setFocusToMenuitem( this.lastMenuitem );
+	}
+
+	setFocusToPreviousMenuitem( currentMenuitem ) {
+		var newMenuitem, index;
+
+		if ( currentMenuitem === this.firstMenuitem ) {
+			newMenuitem = this.lastMenuitem;
+		} else {
+			index = this.menuitemNodes.indexOf( currentMenuitem );
+			newMenuitem = this.menuitemNodes[ index - 1 ];
+		}
+
+		this.setFocusToMenuitem( newMenuitem );
+
+		return newMenuitem;
+	}
+
+	setFocusToNextMenuitem( currentMenuitem ) {
+		var newMenuitem, index;
+
+		if ( currentMenuitem === this.lastMenuitem ) {
+			newMenuitem = this.firstMenuitem;
+		} else {
+			index = this.menuitemNodes.indexOf( currentMenuitem );
+			newMenuitem = this.menuitemNodes[ index + 1 ];
+		}
+		this.setFocusToMenuitem( newMenuitem );
+
+		return newMenuitem;
+	}
+
+	setFocusByFirstCharacter( currentMenuitem, char ) {
+		var start, index;
+
+		if ( char.length > 1 ) {
+			return;
+		}
+
+		char = char.toLowerCase();
+
+		// Get start index for search based on position of currentItem
+		start = this.menuitemNodes.indexOf( currentMenuitem ) + 1;
+		if ( start >= this.menuitemNodes.length ) {
+			start = 0;
+		}
+
+		// Check remaining slots in the menu
+		index = this.firstChars.indexOf( char, start );
+
+		// If not found in remaining slots, check from beginning
+		if ( index === -1 ) {
+			index = this.firstChars.indexOf( char, 0 );
+		}
+
+		// If match was found...
+		if ( index > -1 ) {
+			this.setFocusToMenuitem( this.menuitemNodes[ index ] );
+		}
+	}
+
+	// Utilities
+
+	getIndexFirstChars( startIndex, char ) {
+		for ( var i = startIndex; i < this.firstChars.length; i++ ) {
+			if ( char === this.firstChars[ i ] ) {
+				return i;
+			}
+		}
+		return -1;
+	}
+
+	// Popup menu methods
+
+	openPopup() {
+		this.dropdown.classList.add( "open" );
+		this.buttonNode.setAttribute( "aria-expanded", "true" );
+	}
+
+	closePopup() {
+		if ( this.isOpen() ) {
+			this.buttonNode.removeAttribute( "aria-expanded" );
+			this.dropdown.classList.remove( "open" );
+		}
+	}
+
+	isOpen() {
+		return this.buttonNode.getAttribute( "aria-expanded" ) === "true";
+	}
+
+	// Menu event handlers
+
+	onFocusin() {
+		this.dropdown.classList.add( "focus" );
+	}
+
+	onFocusout() {
+		this.dropdown.classList.remove( "focus" );
+	}
+
+	onButtonKeydown( event ) {
+		var key = event.key,
+			flag = false;
+
+		switch ( key ) {
+		case " ":
+		case "Enter":
+		case "ArrowDown":
```

Comment:
```suggestion
		case "ArrowDown":
```

Not sure if we should open up the dropdown when the user use the arrow key. Some user do read the content by using the arrow key. I think relying on an action key like the "enter" or the "space bar" only to open the dropdown may provide a better user experience for screen reader. This need to be validated with pattern described in the WAI-ARIA practice.

Reply:
This is a non-issue, that behaviour for "up" and "down" are identified in the Aria pattern here: https://www.w3.org/WAI/ARIA/apg/patterns/combobox/#keyboardinteraction